torchvision/csrc/io/image/cpu/decode_jpeg.cpp

```diff
@@ -191,6 +198,54 @@ torch::Tensor decode_jpeg(const torch::Tensor& data, ImageReadMode mode) {
     jpeg_calc_output_dimensions(&cinfo);
   }
 
+  int exif_orientation = 0;
+  if (apply_exif_orientation) {
+    // Check for Exif marker APP1
+    jpeg_saved_marker_ptr exif_marker = 0;
+    jpeg_saved_marker_ptr cmarker = cinfo.marker_list;
+    while (cmarker && exif_marker == 0) {
+      if (cmarker->marker == APP1) {
+        exif_marker = cmarker;
+      }
+      cmarker = cmarker->next;
+    }
+
+    if (exif_marker) {
+      // Code below is inspired from OpenCV
+      // https://github.com/opencv/opencv/blob/097891e311fae1d8354eb092a0fd0171e630d78c/modules/modules/imgcodecs/src/exif.cpp
+
+      // Bytes from Exif size field to the first TIFF header
+      constexpr size_t start_offset = 6;
+      if (exif_marker->data_length > start_offset) {
+        auto* exif_data_ptr = exif_marker->data + start_offset;
+        auto size = exif_marker->data_length - start_offset;
+        std::vector<unsigned char> exif_data_vec(
+            exif_data_ptr, exif_data_ptr + size);
+
+        auto endianness = get_endianness(exif_data_vec);
+
+        // Checking whether Tag Mark (0x002A) correspond to one contained in the
+        // Jpeg file
+        uint16_t tag_mark = get_uint16(exif_data_vec, endianness, 2);
+        if (tag_mark == REQ_EXIF_TAG_MARK) {
+          auto offset = get_uint32(exif_data_vec, endianness, 4);
+          size_t num_entry = get_uint16(exif_data_vec, endianness, offset);
+          offset += 2; // go to start of tag fields
+          constexpr size_t tiff_field_size = 12;
+          for (size_t entry = 0; entry < num_entry; entry++) {
```

Comment:
Is there any risk that this loop can segfault/overflow if the entries are malformed?

Reply:
get_uint16 can return -1 in case offset + required size for uint16 go outside the actual buffer size, so we can catch it here and return earlier.
I hope the code is safe but I can't say 100% sure. I have to investigate a bit more.